File: flutter_version/lib/src/features/news/data/repositories/news_repository_impl.dart
import '../../domain/entities/article.dart';
import '../../domain/repositories/news_repository.dart';
import '../datasources/news_remote_datasource.dart';

class NewsRepositoryImpl implements NewsRepository {
  final NewsRemoteDataSource remoteDataSource;
  NewsRepositoryImpl({required this.remoteDataSource});

  @override
  Future<List<Article>> getTopHeadlines({String country = 'us', String? category, String? q}) async {
    final models = await remoteDataSource.getTopHeadlines(country: country, category: category, q: q);
    return models.map((m) => m as Article).toList();
  }
}

File: swiftui_version/News/News/Features/News/Data/Repositories/NewsRepositoryImpl.swift
import Foundation

/// Concrete implementation of NewsRepository
class NewsRepositoryImpl: NewsRepository {
    private let remoteDataSource: NewsRemoteDataSource
    
    init(remoteDataSource: NewsRemoteDataSource) {
        self.remoteDataSource = remoteDataSource
    }
    
    func getTopHeadlines(
        country: String,
        category: String?,
        query: String?,
        page: Int,
        pageSize: Int
    ) async throws -> ArticlesPage {
        let (articleModels, totalResults) = try await remoteDataSource.getTopHeadlines(
            country: country,
            category: category,
            query: query,
            page: page,
            pageSize: pageSize
        )
        
        let articles = articleModels.map { $0.toDomain() }
        
        return ArticlesPage(
            articles: articles,
            totalResults: totalResults
        )
    }
}
